flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java

```diff
@@ -368,19 +372,34 @@ public boolean deregisterKeySelectionListener(KeySelectionListener<K> listener)
         // collections don't change once started and handles are immutable
         List<ChangelogStateHandle> prevDeltaCopy =
                 new ArrayList<>(changelogStateBackendStateCopy.getRestoredNonMaterialized());
+        long incrementalMaterializeSize = 0L;
         if (delta != null && delta.getStateSize() > 0) {
             prevDeltaCopy.add(delta);
+            incrementalMaterializeSize += delta.getIncrementalStateSize();
         }
 
         if (prevDeltaCopy.isEmpty()
                 && changelogStateBackendStateCopy.getMaterializedSnapshot().isEmpty()) {
             return SnapshotResult.empty();
         } else {
+            List<KeyedStateHandle> materializedSnapshot =
+                    changelogStateBackendStateCopy.getMaterializedSnapshot();
+            for (KeyedStateHandle keyedStateHandle : materializedSnapshot) {
+                if (!lastCompletedHandles.contains(keyedStateHandle)) {
+                    incrementalMaterializeSize += keyedStateHandle.getStateSize();
```

Comment:
Let me expand my concerns directly in this comment:
1. If sending handles fails (or aborted) then UI will show inconsistent sizes (missing aborted info).
2. It requires handles to properly implement hashCode/equals. Otherwise, wrapped backend or writer can construct a new handle pointing to the same files; I think this is currently the case
3. If two checkpoints finish out-of-order and call `buildSnapshotResult` then the size will be incorrect 
4. Incremental size is computed in two places: inside `getIncrementalStateSize`/backend and in `buildSnapshotResult` which seems excessive and fragile
5. After materialization, incremental size will be much bigger and async duration smaller. Wouldn't it be confusing for the users?

Reply:
1. If `materializate-1` happens before `chk-10`, even `chk-10` failed finally, we would still get a large number of incremental state size. And the next succss `chk-11` would be small. I prefer this plan, WDYT?
2. I think we could leverage materialization id to judge the delta state handles and simplify the logic to a map of `<materization id, materizated state size>`
3. If we can use materizaton id to distiguish, the out-of-order problem could be solved.
4. I think this is reasonable as we have two parts here.
5. We can add documentation in changelog part then.

Reply:
I agree with 2, 3, 4, but still have concerns on 1, and more importantly 5.

> 1. even chk-10 failed finally, we would still get a large number of incremental state size

But the materialization state size will be lost, right?
To fix this, we'll have to track **sent** materialization sizes (by materialization and checkpoint ID); mark them as **reported** upon receiving checkpoint confirmation; and then cleanup once next materialization completes.
To me, this complexity seems unjustified.

> 5. We can add documentation in changelog part then.

I think that most users want to know two things regarding checkpoint size:
1. Total size on DFS (i.e. including materialized state)
2. If a checkpoint was slow, how much data was uploaded in the async phase - i.e. **without** materialization; including materialization size would actually do harm here, as it will complicate the reasoning about async phase duration

Reply:
If we do not include the materialization part, we will do not know when the materialization completed on each task via the web UI. And we can explain the large incremental checkpoint size compared with short async duration in documentations of changelog state-backend.

I do think we need to think of it carefully how we define the incremental state size fo changelog state-backend.

If the materialization begins after `chk-9` and completes before `chk-11`. Will `chk-11` include that part? 
Moreover, If we think the incremental checkpoint size is the incremental size compared with last completed checkpoint, what about the case that `chk-11` failed globally but `chk-12` succeed in the end, will `chk-12` include that materialized part?

Maybe we can change the definiation of incremental checkpoint size to "The incremental checkpoint size compared with last checkpoint (not the last completed one)", we can say only `chk-11` would include that materialization part no matter whether it succeed in the end.

By doing so, if we can avoid to reupload SST files in the future, we can also apply the semantics of incremental checkpoint size.

Reply:
Or maybe we can explain the semantics as:

- For general incremental checkpoint, the incremental checkpoint data size means the delta data size comapred with last completed checkpoint.
- For changelog based checkpoint, the incremental checkpoint data size means the delta data size comapred with last triggered checkpoint.

Alothough it looks a bit complex, we can remove the first one if we can support avoiding re-upload for RocksDB incremental checkpoint in the future.

Reply:
> If we do not include the materialization part, we will do not know when the materialization completed on each task via the web UI.
Why do we need to know this? (we can infer by the reduced total size but why?)

I'd propose to derive the definition from the user needs, which I desribed above (maybe you can add or correct?).
If so, "incremental checkpoint size" should be the size of data uploaded during the async phase of the given checkpoint.

This should work for: Changelog, RocksDB with re-upload, RocksDB without re-upload, or any other incremental state backend.

Reply:
> If so, "incremental checkpoint size" should be the size of data uploaded during the async phase of the given checkpoint.

I think "async checkpointed size" instead of "incremental checkpoint size" might be more suitable for your defination.

Reply:
"Async Checkpointed Data Size" is fine for me.

But I think metrics/things from UI users can learn:
- Data Exploding
- When a checkpoint includes a new Materialization
- Materialization size
- changelog sizes from the last complete checkpoint (that can roughly infer restore time)

Reply:
How about "Async Persist Checkpoint Data Size", which only represents the async persist checkpoint data size between the checkpoint trigger and complete.

For the last three questions, I think we need to introduce another two columns to describe. Currently, we can only know materialization happened if the full checkpoint size decreased suddenly.

Reply:
The data uploaded during the async phase is (usually) created during the sync phase. So  "Async Persist Checkpoint Data Size" is not very precise. The current UI does distinguish duration of sync and async phases; also nothing prevents backend from persisting everything during the sync phase.

Something like "Foreground persist data size" would be more precise, but it would confuse non-changelog users I guess. WDYT?

So maybe "Sync/async Persist Checkpoint Data Size"?

Reply:
After discussion with Yuan, we can use two columns, one stay the same as "Checkpointed Data Size", the other would be "Full Checkpoint Data Size". WDYT?

Reply:
"Checkpointed Data Size" still doesn't include materialization, right?
Sounds good to me then :)

Reply:
> "Checkpointed Data Size" still doesn't include materialization

Yes, does not include the materialization part.